# Loading data in pandas



## Learning Objectives
* load CSV files
* load JSON files
* use pd.read_html to extract tables from web pages
* load data from simple APIs 
* load data from a SQL database


Our usual mantra:

In [1]:
import pandas as pd

Recall the ```pd.read_csv``` function that you've seen in previous classes:

In [2]:
menu = pd.read_csv('https://raw.githubusercontent.com/umsi-data-science/si370/master/data/menu.csv') 

In [3]:
menu.head(1)

Category          Item    Serving Size  Calories  Calories from Fat  \
0  Breakfast  Egg McMuffin  4.8 oz (136 g)       300                120   

   Total Fat  Total Fat (% Daily Value)  Saturated Fat  \
0       13.0                         20            5.0   

   Saturated Fat (% Daily Value)  Trans Fat  ...  Carbohydrates  \
0                             25        0.0  ...             31   

   Carbohydrates (% Daily Value)  Dietary Fiber  \
0                             10              4   

   Dietary Fiber (% Daily Value)  Sugars  Protein  Vitamin A (% Daily Value)  \
0                             17       3       17                         10   

   Vitamin C (% Daily Value)  Calcium (% Daily Value)  Iron (% Daily Value)  
0                          0                       25                    15  

[1 rows x 24 columns]

That works great for well-formatted CSV files, but what happens when you get something that looks like the ```data/avocado_eu.csv``` file.

Referring back to your readings and the [read_csv documentation online](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html), complete the following exercise


Let's read the data/avocado_eu.csv file into a pandas DataFrame and show the first 5 rows.


In [4]:
avocado = pd.read_csv('https://raw.githubusercontent.com/umsi-data-science/si370/master/data/avocado_eu.csv')
avocado.head(5)

;Date;AveragePrice;Total Volume;4046;4225;4770;Total Bags;Small Bags;Large Bags;XLarge Bags;type;year;region
0;2015-12-27;1 33;64236  62;1036 74;54454 85;48  16;8696 87;8603 62;93  25;0                         0;conventional;2015;Albany                                                          
1;2015-12-20;1 35;54876  98;674  28;44638 81;58  33;9505 56;9408 07;97  49;0                         0;conventional;2015;Albany                                                          
2;2015-12-13;0 93;118220 22;794  7;109149 67;130 5;8145  35;8042 21;103 14;0                         0;conventional;2015;Albany                                                          
3;2015-12-06;1 08;78992  15;1132 0;71976  41;72  58;5811 16;5677 4;133  76;0                         0;conventional;2015;Albany                                                          
4;2015-11-29;1 28;51039  6;941   48;43838 39;75  78;6183 95;5986 26;197 69;0                         0;conventional;2015;Albany

You'll notice that, unless you did something special in the previous read_csv invocation, the decimal points don't look quite right.  Go ahead and find the right option to convert commas to periods when loading a CSV file, as well as set the index column correctly (without having it duplicated):

## IVQ Exercise:
Read the data/avocado_eu.csv file using the correct delimiter and decimal character into a dataframe and show the first 5 rows:

The output from should look like:

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Date</th>
      <th>AveragePrice</th>
      <th>Total Volume</th>
      <th>4046</th>
      <th>4225</th>
      <th>4770</th>
      <th>Total Bags</th>
      <th>Small Bags</th>
      <th>Large Bags</th>
      <th>XLarge Bags</th>
      <th>type</th>
      <th>year</th>
      <th>region</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>2015-12-27</td>
      <td>1.33</td>
      <td>64236.62</td>
      <td>1036.74</td>
      <td>54454.85</td>
      <td>48.16</td>
      <td>8696.87</td>
      <td>8603.62</td>
      <td>93.25</td>
      <td>0.0</td>
      <td>conventional</td>
      <td>2015</td>
      <td>Albany</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2015-12-20</td>
      <td>1.35</td>
      <td>54876.98</td>
      <td>674.28</td>
      <td>44638.81</td>
      <td>58.33</td>
      <td>9505.56</td>
      <td>9408.07</td>
      <td>97.49</td>
      <td>0.0</td>
      <td>conventional</td>
      <td>2015</td>
      <td>Albany</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2015-12-13</td>
      <td>0.93</td>
      <td>118220.22</td>
      <td>794.70</td>
      <td>109149.67</td>
      <td>130.50</td>
      <td>8145.35</td>
      <td>8042.21</td>
      <td>103.14</td>
      <td>0.0</td>
      <td>conventional</td>
      <td>2015</td>
      <td>Albany</td>
    </tr>
    <tr>
      <th>3</th>
      <td>2015-12-06</td>
      <td>1.08</td>
      <td>78992.15</td>
      <td>1132.00</td>
      <td>71976.41</td>
      <td>72.58</td>
      <td>5811.16</td>
      <td>5677.40</td>
      <td>133.76</td>
      <td>0.0</td>
      <td>conventional</td>
      <td>2015</td>
      <td>Albany</td>
    </tr>
    <tr>
      <th>4</th>
      <td>2015-11-29</td>
      <td>1.28</td>
      <td>51039.60</td>
      <td>941.48</td>
      <td>43838.39</td>
      <td>75.78</td>
      <td>6183.95</td>
      <td>5986.26</td>
      <td>197.69</td>
      <td>0.0</td>
      <td>conventional</td>
      <td>2015</td>
      <td>Albany</td>
    </tr>
  </tbody>
</table>

In [13]:
avocado = pd.read_csv('https://raw.githubusercontent.com/umsi-data-science/si370/master/data/avocado_eu.csv', delimiter=';', decimal=',')
avocado.head(5)

Unnamed: 0        Date  AveragePrice  Total Volume     4046       4225  \
0          0  2015-12-27          1.33      64236.62  1036.74   54454.85   
1          1  2015-12-20          1.35      54876.98   674.28   44638.81   
2          2  2015-12-13          0.93     118220.22   794.70  109149.67   
3          3  2015-12-06          1.08      78992.15  1132.00   71976.41   
4          4  2015-11-29          1.28      51039.60   941.48   43838.39   

     4770  Total Bags  Small Bags  Large Bags  XLarge Bags          type  \
0   48.16     8696.87     8603.62       93.25          0.0  conventional   
1   58.33     9505.56     9408.07       97.49          0.0  conventional   
2  130.50     8145.35     8042.21      103.14          0.0  conventional   
3   72.58     5811.16     5677.40      133.76          0.0  conventional   
4   75.78     6183.95     5986.26      197.69          0.0  conventional   

   year  region  
0  2015  Albany  
1  2015  Albany  
2  2015  Albany  
3  2015  Albany  
4  2015  Albany

# Loading JSON data

In addition to CSV files, JSON (JavaScript Object Notation) files or data are commonly used.  

In [14]:
nfl_football_players = pd.read_json('https://raw.githubusercontent.com/umsi-data-science/si370/master/data/nfl_football_profiles.json')

In [15]:
nfl_football_players.head()

player_id             name position height  weight    current_team  \
0       1809  Robert Blackmon       DB    6-0   208.0            None   
1      23586       Dean Wells       LB    6-3   248.0            None   
2        355      Kiko Alonso      ILB    6-3   238.0  Miami Dolphins   
3      18182     Steve Ramsey       QB    6-2   210.0            None   
4      16250       Cory Nelms       CB    6-0   195.0            None   

   birth_date     birth_place  death_date      college       high_school  \
0  1967-05-12    Bay City, TX        None       Baylor     Van Vleck, TX   
1  1970-07-20  Louisville, KY        None     Kentucky    Holy Cross, KY   
2  1990-08-14      Newton, MA        None       Oregon     Los Gatos, CA   
3  1948-04-22      Dallas, TX  1999-10-15  North Texas  W.W. Samuell, TX   
4  1988-02-27     Neptune, NJ        None   Miami (FL)       Neptune, NJ   

           draft_team  draft_round  draft_position  draft_year current_salary  \
0    Seattle Seahawks          2.0            34.0      1990.0           None   
1    Seattle Seahawks          4.0            85.0      1993.0           None   
2       Buffalo Bills          2.0            46.0      2013.0      1,075,000   
3  New Orleans Saints          5.0           126.0      1970.0           None   
4                None          NaN             NaN         NaN           None   

   hof_induction_year  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN

And, just for fun, show the player with the highest Current Salary from that dataset:

In [17]:
nfl_football_players.sort_values('current_salary', ascending=False).head(1)

player_id               name position height  weight  \
6454        721  Jeremiah Attaochu      OLB    6-3   252.0   

              current_team  birth_date      birth_place death_date  \
6454  Los Angeles Chargers  1993-01-17  Ibadan, Nigeria       None   

           college             high_school          draft_team  draft_round  \
6454  Georgia Tech  Archbishop Carroll, DE  San Diego Chargers          2.0   

      draft_position  draft_year current_salary  hof_induction_year  
6454            50.0      2014.0        993,150                 NaN

## Fixing up the data
Assuming you did something like sort_values on one of the original columns, you probably got the wrong result.

Looking a bit more closely at the results, you'll notice that the current_salary column.  Remembering that we have made the shift from pythonic to pandorable, we can leverage the impressive-sounding "vectorized string functions" mentioned in the McKinney book.  Specifically, we can use the str.replace(...) method.  Note that had we use read_csv to load the file we could have used the ```thousands=``` option and avoided all this, but sometimes data doesn't come in a convenient format.

One way to apply functions is to operate on a column and then assign the results to another column.  For example, if we wanted to eliminate commas, we could replace them with null strings

In [28]:
nfl_football_players['current_salary']
# we can find that this series dtype == object (most of time is str, so can not sort right

0             None
1             None
2        1,075,000
3             None
4             None
           ...    
25038         None
25039         None
25040         None
25041         None
25042         None
Name: current_salary, Length: 25043, dtype: object

In [29]:
# str.replace() to replace ','
nfl_football_players['current_salary'].str.replace(',', '')

0           None
1           None
2        1075000
3           None
4           None
          ...   
25038       None
25039       None
25040       None
25041       None
25042       None
Name: current_salary, Length: 25043, dtype: object

And assign the results to a column in the original dataframe (in this case I'm calling the column current_salary_nocommas)

In [31]:
# assign to new column
nfl_football_players['current_salary_nocommas'] = nfl_football_players['current_salary'].str.replace(',', '')

But you'll notice that the type of the column is string, and we want to convert it to a float so we can sort it numerically.  So we can use the astype() function to convert it:

In [39]:
# look up dtype
nfl_football_players.current_salary_nocommas.dtype

dtype('O')

In [ ]:
# dtype('O') means 'object' (str)

In [35]:
# .astype(float) convert str to float 
nfl_football_players['current_salary_cleaned'] = nfl_football_players['current_salary_nocommas'].astype(float)

In [36]:
nfl_football_players.head(3)

player_id             name position height  weight    current_team  \
0       1809  Robert Blackmon       DB    6-0   208.0            None   
1      23586       Dean Wells       LB    6-3   248.0            None   
2        355      Kiko Alonso      ILB    6-3   238.0  Miami Dolphins   

   birth_date     birth_place death_date   college     high_school  \
0  1967-05-12    Bay City, TX       None    Baylor   Van Vleck, TX   
1  1970-07-20  Louisville, KY       None  Kentucky  Holy Cross, KY   
2  1990-08-14      Newton, MA       None    Oregon   Los Gatos, CA   

         draft_team  draft_round  draft_position  draft_year current_salary  \
0  Seattle Seahawks          2.0            34.0      1990.0           None   
1  Seattle Seahawks          4.0            85.0      1993.0           None   
2     Buffalo Bills          2.0            46.0      2013.0      1,075,000   

   hof_induction_year current_salary_nocommas  current_salary_cleaned  
0                 NaN                    None                     NaN  
1                 NaN                    None                     NaN  
2                 NaN                 1075000               1075000.0

And now we can re-run our command to sort by salary and get the correct result:

In [38]:
nfl_football_players.sort_values('current_salary_cleaned', ascending=False).head(1)

player_id          name position height  weight         current_team  \
17756       4644  Kirk Cousins       QB    6-3   214.0  Washington Redskins   

       birth_date  birth_place death_date       college  \
17756  1988-08-19  Holland, MI       None  Michigan St.   

                 high_school           draft_team  draft_round  \
17756  Holland Christian, MI  Washington Redskins          4.0   

       draft_position  draft_year current_salary  hof_induction_year  \
17756           102.0      2012.0     23,943,600                 NaN   

      current_salary_nocommas  current_salary_cleaned  
17756                23943600              23943600.0

## Dropping missing values

In addition to the "all" or "any" functionality described in McKinney section 7.1, it's sometimes useful to drop a row only if a certain column or columns have missing data.  To do this, use the subset= option with dropna().  So, for example, to drop all players for whom we do not have salary information, we could use the following code:

In [42]:
# .dropna(subset=[]) drop when subset has NA value
nfl_football_players_salaries = nfl_football_players.dropna(subset=['current_salary_cleaned'])

In [45]:
nfl_football_players_salaries.head()

player_id               name position height  weight      current_team  \
2         355        Kiko Alonso      ILB    6-3   238.0    Miami Dolphins   
6        2701      Preston Brown      ILB    6-1   251.0     Buffalo Bills   
13       3966        Frank Clark       DE    6-2   270.0  Seattle Seahawks   
37      19449  Emmanuel Sanders        WR   5-11   186.0    Denver Broncos   
53      17858       Bilal Powell       RB   5-11   204.0     New York Jets   

    birth_date     birth_place death_date     college      high_school  \
2   1990-08-14      Newton, MA       None      Oregon    Los Gatos, CA   
6   1992-10-27  Cincinnati, OH       None  Louisville    Northwest, OH   
13  1993-06-14   Cleveland, OH       None    Michigan    Glenville, OH   
37  1987-03-16   Bellville, TX       None         SMU    Bellville, TX   
53  1988-10-27    Lakeland, FL       None  Louisville  Lake Gibson, FL   

             draft_team  draft_round  draft_position  draft_year  \
2         Buffalo Bills          2.0            46.0      2013.0   
6         Buffalo Bills          3.0            73.0      2014.0   
13     Seattle Seahawks          2.0            63.0      2015.0   
37  Pittsburgh Steelers          3.0            82.0      2010.0   
53        New York Jets          4.0           126.0      2011.0   

   current_salary  hof_induction_year current_salary_nocommas  \
2       1,075,000                 NaN                 1075000   
6       1,762,000                 NaN                 1762000   
13        774,294                 NaN                  774294   
37      6,750,000                 NaN                 6750000   
53      3,750,000                 NaN                 3750000   

    current_salary_cleaned  
2                1075000.0  
6                1762000.0  
13                774294.0  
37               6750000.0  
53               3750000.0

In [ ]:
# # .dropna()  will drop rows has NA value in any column

# Scraping Tables from HTML

The ```pd.read_html``` function returns a list of DataFrames read from an HTML source.  The following line will return a list of DataFrames from https://en.wikipedia.org/wiki/List_of_largest_sports_contracts

In [47]:
# certificate
import ssl
ssl._create_default_https_context = ssl._create_unverified_context

In [52]:
# pd.read_html() will extract all tables in the page
contracts_scraped = pd.read_html('https://en.wikipedia.org/wiki/List_of_largest_sports_contracts',header=0)

In [53]:
# contracts_scraped is a list of dataframes, so len() will get how many df we get
len(contracts_scraped)

1

To get the first table, you'll need to pull off the 0th element:

In [54]:
# get the first df/element
contracts = contracts_scraped[0]
contracts.head()

Rank                 Name         Organization                 Sport  \
0    1     Lionel Messi[b]‡        FC Barcelona*  Association football   
1    2  Patrick Mahomes[d]‡   Kansas City Chiefs     American football   
2    3           Mike Trout   Los Angeles Angels              Baseball   
3    4       Canelo Álvarez                DAZN*                Boxing   
4    5         Mookie Betts  Los Angeles Dodgers              Baseball   

     Length of contract Contract value (USD) Average per year (USD)  \
0   4 years (2017–2021)         $673,919,105           $168,479,776   
1  10 years (2020–2030)         $503,000,000            $41,916,667   
2  12 years (2019–2030)         $426,500,000            $35,541,667   
3   5 years (2018–2023)         $365,000,000            $73,000,000   
4  12 years (2020–2032)         $365,000,000            $30,416,667   

  Average per game/event[a] (USD)  Ref  
0                   $3,743,995[c]  [1]  
1                   $2,465,686[e]  [2]  
2                        $219,393  [3]  
3                     $33,181,818  [4]  
4                        $187,757  [5]

## IVQ Exercise: 

Count the number of players from each sport in the List of Largest Sports Contracts 

Hint:  remember ```value_counts()``` from your previous work.

In [56]:
contracts['Sport'].value_counts()

Baseball                52
Basketball              37
American football        7
Association football     2
Boxing                   1
Auto racing              1
Name: Sport, dtype: int64

# APIs and requests
You've covered the ```requests``` package in previous courses.  This example shows what you can do with an API that returns JSON:

In [57]:
import requests

In [58]:
url = 'https://api.github.com/repos/pandas-dev/pandas/issues'

In [59]:
# requests.get(), get url
resp = requests.get(url)

In [62]:
# to json, assign to 
data = resp.json()

In [64]:
#data

In [65]:
data[0]['user']['login'] 

'jbrockmendel'

In [66]:
# initialize a dataframe with that data we get
issues = pd.DataFrame(data)
issues.head()

url  \
0  https://api.github.com/repos/pandas-dev/pandas...   
1  https://api.github.com/repos/pandas-dev/pandas...   
2  https://api.github.com/repos/pandas-dev/pandas...   
3  https://api.github.com/repos/pandas-dev/pandas...   
4  https://api.github.com/repos/pandas-dev/pandas...   

                                   repository_url  \
0  https://api.github.com/repos/pandas-dev/pandas   
1  https://api.github.com/repos/pandas-dev/pandas   
2  https://api.github.com/repos/pandas-dev/pandas   
3  https://api.github.com/repos/pandas-dev/pandas   
4  https://api.github.com/repos/pandas-dev/pandas   

                                          labels_url  \
0  https://api.github.com/repos/pandas-dev/pandas...   
1  https://api.github.com/repos/pandas-dev/pandas...   
2  https://api.github.com/repos/pandas-dev/pandas...   
3  https://api.github.com/repos/pandas-dev/pandas...   
4  https://api.github.com/repos/pandas-dev/pandas...   

                                        comments_url  \
0  https://api.github.com/repos/pandas-dev/pandas...   
1  https://api.github.com/repos/pandas-dev/pandas...   
2  https://api.github.com/repos/pandas-dev/pandas...   
3  https://api.github.com/repos/pandas-dev/pandas...   
4  https://api.github.com/repos/pandas-dev/pandas...   

                                          events_url  \
0  https://api.github.com/repos/pandas-dev/pandas...   
1  https://api.github.com/repos/pandas-dev/pandas...   
2  https://api.github.com/repos/pandas-dev/pandas...   
3  https://api.github.com/repos/pandas-dev/pandas...   
4  https://api.github.com/repos/pandas-dev/pandas...   

                                          html_url          id  \
0  https://github.com/pandas-dev/pandas/pull/45672  1116940190   
1  https://github.com/pandas-dev/pandas/pull/45671  1116921040   
2  https://github.com/pandas-dev/pandas/pull/45670  1116897054   
3  https://github.com/pandas-dev/pandas/pull/45669  1116876778   
4  https://github.com/pandas-dev/pandas/pull/45668  1116874482   

               node_id  number  \
0  PR_kwDOAA0YD84xtpmY   45672   
1  PR_kwDOAA0YD84xtlwk   45671   
2  PR_kwDOAA0YD84xtg-n   45670   
3  PR_kwDOAA0YD84xtcof   45669   
4  PR_kwDOAA0YD84xtcJI   45668   

                                               title  ...  \
0                   REF: implement LossySetitemError  ...   
1  Backport PR #45565 on branch 1.4.x (REGR: chec...  ...   
2                         Add DataFrame.iterrowdicts  ...   
3  Backport PR #45612 on branch 1.4.x (DOC: Impro...  ...   
4      BUG: unnecessary FutureWarning in sort_values  ...   

             updated_at closed_at author_association  active_lock_reason  \
0  2022-01-28T02:38:52Z      None             MEMBER                None   
1  2022-01-28T02:00:02Z      None               NONE                None   
2  2022-01-28T02:35:49Z      None               NONE                None   
3  2022-01-28T00:34:47Z      None               NONE                None   
4  2022-01-28T02:30:57Z      None             MEMBER                None   

   draft                                       pull_request  \
0  False  {'url': 'https://api.github.com/repos/pandas-d...   
1  False  {'url': 'https://api.github.com/repos/pandas-d...   
2  False  {'url': 'https://api.github.com/repos/pandas-d...   
3  False  {'url': 'https://api.github.com/repos/pandas-d...   
4  False  {'url': 'https://api.github.com/repos/pandas-d...   

                                                body  \
0  - [ ] closes #xxxx\r\n- [ ] tests added / pass...   
1  Backport PR #45565: REGR: check_flags not resp...   
2  - [x] tests added / passed\r\n- [ ] Ensure all...   
3     Backport PR #45612: DOC: Improve reshaping.rst   
4  - [x] closes #45618\r\n- [x] tests added / pas...   

                                           reactions  \
0  {'url': 'https://api.github.com/repos/pandas-d...   
1  {'url': 'https://api.github.com/repos/pandas-d...   
2  {'url': 'https://api.github.com/repos/pandas-d...   
3  {'url': 'http

In [68]:
# columns
issues.columns

Index(['url', 'repository_url', 'labels_url', 'comments_url', 'events_url',
       'html_url', 'id', 'node_id', 'number', 'title', 'user', 'labels',
       'state', 'locked', 'assignee', 'assignees', 'milestone', 'comments',
       'created_at', 'updated_at', 'closed_at', 'author_association',
       'active_lock_reason', 'draft', 'pull_request', 'body', 'reactions',
       'timeline_url', 'performed_via_github_app'],
      dtype='object')

In [70]:
# get df within the columns we want (subset)
issues = pd.DataFrame(data, columns=['number', 'title','labels', 'state'])
issues.head()

number                                              title  \
0   45672                   REF: implement LossySetitemError   
1   45671  Backport PR #45565 on branch 1.4.x (REGR: chec...   
2   45670                         Add DataFrame.iterrowdicts   
3   45669  Backport PR #45612 on branch 1.4.x (DOC: Impro...   
4   45668      BUG: unnecessary FutureWarning in sort_values   

                                              labels state  
0                                                 []  open  
1  [{'id': 127685, 'node_id': 'MDU6TGFiZWwxMjc2OD...  open  
2                                                 []  open  
3  [{'id': 134699, 'node_id': 'MDU6TGFiZWwxMzQ2OT...  open  
4  [{'id': 49182326, 'node_id': 'MDU6TGFiZWw0OTE4...  open

In [72]:
# iloc[] get the 0th row of df
issues.iloc[0]

number                               45672
title     REF: implement LossySetitemError
labels                                  []
state                                 open
Name: 0, dtype: object

# Accessing databases
This section requires you to review your notes about SQL from previous courses.

We'll use this opportunity to gauge where we're at in terms of understanding SQL.

---



Let's install sqlalchemy and some database drivers.
(Does that make sense?  If not, let's talk about it when we get together.)

In [73]:
!pip install sqlalchemy pymysql

And import the relevant libraries:

In [74]:
import sqlalchemy as sqla
import pandas as pd

Now let's set up some variables and connect to the database.  Note that the password is available through the video and you'll need to replace SECRET_REPLACE_ME with the password.

In [75]:
# connect to the database
endpoint = "database-370-fa2020.cgeyvh9t1svz.us-east-1.rds.amazonaws.com" 
username = "si370wn2020"
password = "iheartdatascience"
database = "goodreads"

db = sqla.create_engine(f'mysql+pymysql://{username}:{password}@{endpoint}/{database}')

Now the real part:  using `read_sql` to read some data from the database.  In this case, we're going to count the number of entries in the ratings table:

In [ ]:
# read database
df = pd.read_sql('select count(*) from ratings',db)
df.head()

We can get the SQL engine to do more of the work (instead of doing it in pandas):

In [ ]:
df = pd.read_sql('select rating, count(*) from ratings group by rating', db)
df.head()
# different sql queries 

## Summary
* load CSV files using `read_csv()` (options are useful)
* load JSON files using `read_json()`
* extract tables from web pages using `read_html()`
* load data from simple APIs by initializing a DataFrame from JSON (as an example)
* load data from a SQL database using `read_sql()`